pkg/input/arm/eval.go

```diff
@@ -0,0 +1,82 @@
+// © 2022-2023 Snyk Limited All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arm
+
+import (
+	"strings"
+)
+
+// Makes state available to ARM template function evaluation. It contains
+// references to implementations of functions, and the other fields exist to
+// pass data to certain function impls other than what is already in their args.
+type EvaluationContext struct {
+	discoveredResourceSet map[string]struct{}
+	funcs                 map[string]armFnImpl
+}
+
+func NewEvaluationContext(discoveredResourceSet map[string]struct{}) *EvaluationContext {
+	evalCtx := &EvaluationContext{
+		discoveredResourceSet: discoveredResourceSet,
+		funcs: map[string]armFnImpl{
+			"concat":        concatImpl,
+			"resourceGroup": resourceGroupImpl,
+		},
+	}
+	evalCtx.funcs["resourceId"] = evalCtx.resourceIDImpl
+	return evalCtx
+}
+
+// Detects whether an ARM string is an expression (enclosed by []), and
+// tries to evaluate it if so.
+func (e *EvaluationContext) EvaluateARMTemplateString(input string) (interface{}, error) {
+	if !isARMTemplateExpression(input) {
+		return input, nil
+	}
+
+	return e.eval(extractAndEscapeARMTemplate(input))
```

Comment:
This is a weakly-held opinion, because in all my years working on Go projects I still end up being slightly unsure how to handle errors: returning a stdlib `error` feels "more idiomatic" and flexible, because we don't have to crowbar all cases into our domain-specific error type if we don't want to, and the cost is just another case in a type switch/assertion. I know it's a bit theoretical since atm all of our error paths _do_ return our own error type, but still...

Minor point, but we'd have to return the pointer rather than value, to be able to check against nil.